tests/components/slack/test_notify.py

```diff
@@ -29,84 +36,397 @@
+            mock_response
+        )
+
+        service = SlackNotificationService(hass_mock, mock_client, CONF_DATA)
+
+        await service.async_send_message(
+            "test",
+            target=["#general", "U123456789"],
+            data={"file": {"url": "http://example.com/image.jpg"}},
+        )
+
+        # Should resolve targets and call files_upload_v2 once with all resolved channel IDs
+        assert mock_client.conversations_list.call_count == 2
+        assert mock_client.conversations_open.call_count == 1
+        assert mock_client.files_upload_v2.call_count == 2
+
+
+async def test_remote_file_not_allowed(hass_mock, mock_client) -> None:
+    """Test uploading a remote file from a non-allowed URL."""
+    hass_mock.config.is_allowed_external_url = Mock(return_value=False)
+    service = SlackNotificationService(hass_mock, mock_client, CONF_DATA)
+
+    await service.async_send_message(
+        "test", data={"file": {"url": "http://example.com/not-allowed.jpg"}}
+    )
+
+    mock_client.files_upload_v2.assert_not_called()
+
+
+async def test_remote_file_download_error(mock_client, hass_mock) -> None:
+    """Test handling of remote file download errors."""
+    with patch(
+        "homeassistant.components.slack.notify.aiohttp_client.async_get_clientsession"
+    ) as mock_session:
+        mock_session.return_value.get.side_effect = ClientError()
+
+        service = SlackNotificationService(hass_mock, mock_client, CONF_DATA)
+
+        await service.async_send_message(
+            "test", data={"file": {"url": "http://example.com/error.jpg"}}
+        )
+
+        mock_client.files_upload_v2.assert_not_called()
+
+
+async def test_slack_api_errors(mock_client) -> None:
+    """Test handling of Slack API errors."""
+    mock_client.chat_postMessage.side_effect = SlackApiError("error", {"ok": False})
+    service = SlackNotificationService(None, mock_client, CONF_DATA)
+
+    # Should not raise an exception
+    await service.async_send_message("test", target="C123456789")
+
+
+async def test_conversation_list_error(mock_client) -> None:
+    """Test handling of conversation list errors."""
+    mock_client.conversations_list.side_effect = SlackApiError("error", {"ok": False})
+    service = SlackNotificationService(None, mock_client, CONF_DATA)
+
+    with pytest.raises(HomeAssistantError) as excinfo:
+        await service.async_send_message("test", target="#some-channel")
+
+    assert excinfo.value.translation_domain == DOMAIN
+    assert excinfo.value.translation_key == "error_getting_channel"
+    assert excinfo.value.translation_placeholders == {
+        "channel_name": "some-channel",
+        "error": "error\nThe server responded with: {'ok': False}",
+    }
+
+    mock_client.chat_postMessage.assert_not_called()
+
+
+async def test_channel_not_found(mock_client) -> None:
+    """Test handling of non-existent channel."""
+    service = SlackNotificationService(None, mock_client, CONF_DATA)
```

Comment:
Usually we want to setup the service and test this through the normal APIs instead of using the object directly.